### DISCUSSION TITLE
Kanes's method, velocity constraints

### DISCUSSION THREAD

**Discussion by Peter230655** [sympy/sympy#21718] [Q&A]
Sympy Kane's method allows the use of velocity constraints, say u_j = f(u_k), where u_k is independent.

My question: 
When I numerically integrate the equations of motion, I have to give initial conditions. For u_j(0), do I have to give a value, which satisfies the velocity constraint, or will my initial condition for u_j be ignored and calculated from u_k via u_j(0) = f(u_k(0)) ?

Thanks for any help!

**Comment by moorepants** [comment]
Yes, you would have to give initial conditions that satisfy the constraint. In general, the user is responsible for providing valid initial conditions for the system whether constrained or not.

And if you haven't reduced the equations of motion fully to the minimal set (# DoF) then you may also need to use a DAE integration routine to prevent drift during integration.

**Reply by Peter230655** [reply]
Clear & thanks a lot!
What does #DoF mean? 
What does DAE mean?
I guess, by drift you mean, that the velocity constraints are not always fulfilled 100%, not even 'on average', but may drift in some direction.

**Reply by Peter230655** [reply]
In one problem, I did some time ago, I used a velocity constraint for a holonomic constraint, which cannot be used in Kane's equations. This worked very well, the generalized coordinate, whose velocity I forced to be zero, was nearly constant - as I had hoped.
Thanks again!

**Comment by moorepants** [comment]
`#` DoF : number of degrees of freedom

DAE : differential algebraic equation

yes, on drift the constraints will not be satisfied over time and the error grows

yes, you can differentiate the holonomic constraint and use it as a velocity constraint, but you will see that the holonomic constraint will drift during integration

**Reply by Peter230655** [reply]
Now all clear, thanks a lot!

**Comment by Peter230655** [comment]
What I calculate is an n particle pendulum.
To get the reaction forces at the fixed point (say, first particle), I use:
reaction_forces = KM.auxiliary_eqs.
Reaction_forces contains the second derivatives of the generalized
coordinates, which seems to make sense mechanically.
I substitute them with the corresponding terms of rhs = KM.rhs(). This
seems to work fine.

Now, if I add a velocity constraint, say, the last particle may only move
up and down, I get the ‚auxiliary speeds‘ and their derivatives in rhs.
I simply set them to zero by substitution - and I get some results.
I have no idea, what is the correct way to get rid of them.

Any help is highly appreciated!




On Wed 7. Jul 2021 at 15:51 Jason K. Moore ***@***.***> wrote:

> DoF : number of degrees of freedom
>
> DAE : differential algebraic equation
>
> yes, on drift the constraints will not be satisfied over time and the
> error grows
>
> yes, you can differentiate the holonomic constraint and use it as a
> velocity constraint, but you will see that the holonomic constraint will
> drift during integration
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/discussions/21718#discussioncomment-974992>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT5MQUUUE7QFZY55XGH6TIDTWRLXJANCNFSM476PSRFQ>
> .
>
-- 
Best regards,

Peter Stahlecker

**Reply by moorepants** [reply]
> I have no idea, what is the correct way to get rid of them.

If you introduce the auxiliary speeds, after you form the equations of motion you set the speeds and their derivatives to zero to get rid of them.

Here is an example: https://nbviewer.jupyter.org/github/moorepants/mae223/blob/master/content/lecture-notebooks/mae223-l19-01.ipynb

This should work whether you have velocity constraints or not.

Here is the Whipple-Carvallo bicycle example but with auxiliary speeds to get the contact forces at the velocity constraints:

https://github.com/moorepants/pydy/blob/bicycle-tire-constraint-forces/examples/bicycle_wheel_contact_constraint_forces/whipple.py
